### ISSUE TITLE
Support lifting programs, especially those that don't fit within a week

### ISSUE THREAD

**Issue by ChemiKyle** [wger-project/wger#848]
## Use case

Programs whose workouts don't fit within a single week schedule are tedious to configure.  
Programs like Stronglifts and Starting Strength are simple enough to just deal with the tedium, but programs like GZCL and 531 - which follow an ABCD cyclic pattern of workouts - but are intended to use a 3 day schedule require 4 individual workout weeks to be made in this example, despite only having 4 unique days.

## Proposal

Ideally, one should be able to specify N workouts cycling over M days (specified weekdays is simplest, though I imagine some PPL routines may deviate from MWF).

Again, using GZCL as an example, there are 4 unique workout weeks: ABC, DAB, CDA, BCD. Ideally there would be an interface much like the "Workout" creation that allows defining an arbitrary number of training days, perhaps assigning them an integer instead of a weekday. The next step would be to pick the days of the week to schedule.

This could fall under a new option, "Training programs".

If it needed to fit within the existing framework, a "training program" could be an advanced option while creating a schedule; the requisite number of "workouts" could be autogenerated until workout "A" appears again on the first scheduled day[^1]. Each of these workouts would then be added in to a looped schedule, each workout's "Weeks" set to 1.[^2]

## Consideration of full periodization support

Periodization support would be necessary to really ship a "Training programs" feature but may be beyond the scope of this issue.

Non-responsive undulation can be handled by the user, but even simple progressive overload is dependent on user logging.  
When initiating a log on a known workout, the previous log for the same workout would be checked if it matched the specified number of reps and requisite weight increase for all sets of the specified exercise. If it did not, the autofilled values would fall back to the next stopping point. An issue here is the programs I've used as examples increase set count the next exercise after a failure, managing this on the frontend may be as simple as generating an additional `<div class ="form-row form-row">` on the log page, but I haven't dug into the backend much.

[^1]: I'm sure this can be pre-calculated, there must be some formula for a set of N ordered elements mapping to a set of M ordered elements having P permutations, but my combinatorics knowledge has left its pigeonhole.

[^2]: I'm poking around the database (much easier thanks to the docker setup!) trying to implement this, albeit outside wger itself. I do this sort of thing for a living in R; I'll try to keep my script general if anyone might find it useful. I'm not familiar with django, but may be able to translate to python3 with standard lib (might need pandas) if that'd be helpful to the dev team.

**Comment by manfre** [comment]
For adaptive programs, the weight and/or reps can vary based upon the previous workout of that type. E.g. With SBS there is a percentage change to the training max (up or down) based upon how many reps in the final set compared to the target number of reps. For 531, you increase training max only if you hit a minimum number or reps in the 1+ AMRAP set.

If there is a singular way of adding training sessions, these adaptive programs could potentially leverage that portion of the API in a pluggable way that essentially does an "add the next scheduled session". This is something that I need in a training program to accommodate my training not being bound to a weekly schedule. It cycles between 5 unique training days over a period of 8-12 days (depends on schedule and how I'm feeling).

I found wger today and am planning to get more familiar with the code to better refine my idea of how the above can fit in to the project. (Really happy to have found a Django based fitness project while determining if I needed to create one)

**Comment by RlndVt** [comment]
> ## Proposal
> 
> [...] Ideally there would be an interface much like the "Workout" creation that allows defining an arbitrary number of training days, perhaps assigning them an integer instead of a weekday. **The next step would be to pick the days of the week to schedule.**

Emphasis mine.

Is it necessary to loop back to the day of the week? I would say it is preferable to just keep it ABCD or 1 2 3 4; and show which is the next workout. 

Using a rigid 'days of the week' of the week system has a few issues: 

1) For people with a more chaotic out-of-gym life; one week they might fit in one workout and another week four, and never on the same day of the week. 
2) As OP mentioned, some lifting programs don't 'fit' in a week. If someone wants to use a ABCD schedule, but train thrice a week we get OP's scenario. But this is the case for any mismatch between scheduling days, and workout days (per week).
3) Missing a workout-day, should not equal missing a workout. e.g. Running a U/L or PPL scheme and missing one day would mean you skip a targeted area. And we all know you should never skip leg-day.

So in conclusion, I would suggest stepping away from days of the week for creating a program.

**Comment by rolandgeider** [comment]
I would also tend to specify how often / in which order / etc. one plans to do the training and then calculating the days of the week on the fly as some kind of nice to have info for the calendar, personal planning, etc. This would have the advantage that if you miss one day the next dates are still right, but we might also want to show a warning or something similar.

In any case, after finishing the exercise rewamp (whenever that might be), this is probably the next bigger feature I'd want to tackle

**Comment by ChemiKyle** [comment]
I think templates are a good idea as long as customizability remains. Perhaps an import/export feature would be a good addition here. Most programs are distributed via spreadsheets with macros which leaves a lot to be desired.

My first thought would be to store program data as JSON - this may not play nice with base SQLite, but there is an extension - for customizable progression based on performance. That opens up quite a can of worms, but I tossed a mockup down at the bottom.

---

> Is it necessary to loop back to the day of the week? I would say it is preferable to just keep it ABCD or 1 2 3 4; and show which is the next workout.

> I would also tend to specify how often / in which order / etc. one plans to do the training and then calculating the days of the week on the fly as some kind of nice to have info for the calendar, personal planning, etc

Indeed, auto calculating this may be best, perhaps setting an acceptable range of rest days and intended sessions per week? I definitely agree that if you miss a day it should not mean skipping that exercise.  
Personally, knowing what exercise, weight, and reps I have to do next is what I'd like to know at a glance. Timestamps don't seem important outside of data entry.

---
Here's a bit of rough JSON off the top of my head for how a cookie-cutter GZCL "A" workout may look; it is perhaps over engineered - and overfit to linear progression GZCL - but I tried to imagine how undulating periodization and AMRAP varied progression may be handled generically.


<details>
<summary>Expand JSON mockup</summary>

In prose, if you hit your rep/set goals, you increase weight by `weight_increment` for the next session of that workout.  
If you miss your reps for any set, your weight remains the same and you fall down to the next `scheme` for the next session of that workout (e.g. if I'm supposed to do 5x3 and I get 3, 3, 3, 2, 1, next session I do 6x2 at the same weight).   
Once you fail the last scheme, your next session is spent trying to find your 3 rep max.

- The final set is AMRAP, if you double your goal reps, you also double your progression.
  - I think 531 the `scheme:weight_increment` would be 0, since progression is solely based on AMRAP.

```json
[
  {
    // workout day is an array of objects representing lifts and metadata
    "A": [
      // Tier 1 lift: low rep, high weight
      {
        "lift_name": "back_squat",
        "scheme": [
          // schemes are ordinal, if you fail progression criteria, you drop to the next scheme
          {
            "sets": 5,
            "reps": 3,
            // successfully complete this workout scheme this many times to enable weight increment,
            // possibly useful for programs that don't progress each sequential session
            "success_limit": 1,
            "weight_increment": 10.0
          },
          {
            "sets": 6,
            "reps": 2,
            "weight_increment": 10.0
          },
          {
            "sets": 10,
            "reps": 1,
            "weight_increment": 10.0
          },
          // after failing all the above schemes, GZCL has you retest to find your 3RM
        ],
        "progression_changes": {
          [
            {
              "scheme": -1,
              "reps": NULL,
              "weight_increment": NULL
            },
            {
              // (optional) success condition, hit double your rep target to increase rate of progression
              "set": "last",
              "reps": "double",
              // choose one of these, flat replacement of usual or multiplier on usual
              "weight_increment": 20.0,
              "bonus_weight_increment_multiplier": 2.0
            },
            {
              // (optional) failure condition, fail to meet reps on any set and you will fall to the next scheme
              "set": "any",
              "reps": "less",
              // optional smaller weight increment?
              "bonus_weight_increment": 0.0
            },
          ]
        }
      },
      // Tier 2 lift: high rep, low weight
      {
        "lift_name": "bench_press",
        "scheme": [
          {
            "sets": 3,
            "reps": 10,
            "weight_increment": 5.0
          },
          {
            "sets": 3,
            "reps": 8,
            "weight_increment": 5.0
          },
          {
            "sets": 3,
            "reps": 8,
            "weight_increment": 5.0
          }
          // after failing the final scheme you simply start back from the top at a lower weight
        ]
        "progression_changes": {
          [
            {
              // failure condition, fail to meet reps on any set and you will fall to the next scheme
              // fail to meet reps on the final scheme and you lower weight and restart
              "set": "any",
              "reps": "less",
              "scheme": -1, // end of scheme array
              // decrease
              "bonus_weight_increment": -20.0
            }
          ]
        }
      }
    ]
```

</details>

**Comment by BeatLink** [comment]
Here is my spin on how this would be implemented. 

Workouts would be organized in two stages: Sessions and Schedules. 

Sessions would be a group of exercises along with all of their details (reps, weights, sets, etc). Eg Push, Pull, Legs or Workout A and Workout B

Schedules would be groups of sessions organized so that sessions can be added to particular days of particular weeks.

Hopefully the below hierarchy helps explain somewhat

- Weeks (eg. Week 1, Week 2, etc)
    - Days (eg. Monday, Wednesday, etc) 
       - Sessions (eg, Push, Pull)

It is essentially the existing system but rather than breaking each workout into particular days, it would be the schedule that would be broken up into days after the weeks. Workouts would just be self contained sessions, devoid of scheduling information.

So for example, a user may create 4 seperate sessions: Push, Pull, Legs and Cardio. 
The user could then combine these sessions to form a schedule, so for example, 

Week 1
Monday - Push, Cardio
Tuesday - Cardio
Wednesday - Pull, Cardio
Thursday - Cardio
Friday - Legs, Cardio

Another Example would be an alternating programme

In this case, the user would create 2 sessions, Workouts A and B. Then it could be implemented as follows:

Week 1 
Monday - A
Tuesday
Wednesday - B
Thursday
Friday - A

Week 2
Monday - B
Tuesday
Wednesday - A
Thursday
Friday - B

Loop Weeks: On

**Comment by BeatLink** [comment]
Hi. I wanted to find out if there was any progress with this design change. Perhaps we could get a poll for the various ideas?

**Comment by rolandgeider** [comment]
Nothing spefic yet, but you are right that we should probably start thinking about specific things to do.

Somebody in another issue suggested taking some inspiration from https://www.liftosaur.com/ about handling progression

**Comment by BeatLink** [comment]
I took a look at that app before and it would be awesome if this could be integrated into wger! 

The first step I think though would be decoupling the workouts from specific weekdays though. Once you have self contained workouts decoupled from scheduling information, you could then look at integrating the scripting syntax from liftosaur.

Any thoughts on the structure I proposed here https://github.com/wger-project/wger/issues/848#issuecomment-1067492323?

**Comment by IsotoxalDev** [comment]
This!! This is something i need as well

**Comment by BeatLink** [comment]
I've created an entity relation diagram to detail my suggested structure. Maybe it could give some guidance with refactoring the models:


![irontech-entity-diagram drawio](https://github.com/wger-project/wger/assets/47278150/b99077fa-bdad-4f19-9dad-afd8cbf26714)


Its still a work in progress but if you need help explaining any part let me know

**Comment by rolandgeider** [comment]
something like that, yeah. Something that we still need to consider for the data model is handling progression, e.g. like in the JSON on [this comment](https://github.com/wger-project/wger/issues/848#issuecomment-1004438669)

**Comment by lediable** [comment]
Any update on this feature?

**Comment by RlndVt** [comment]
> I've created an entity relation diagram to detail my suggested structure. Maybe it could give some guidance with refactoring the models:
> 
> ![irontech-entity-diagram drawio](https://private-user-images.githubusercontent.com/47278150/251435808-b99077fa-bdad-4f19-9dad-afd8cbf26714.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIyMDM3ODYsIm5iZiI6MTcwMjIwMzQ4NiwicGF0aCI6Ii80NzI3ODE1MC8yNTE0MzU4MDgtYjk5MDc3ZmEtYmRhZC00ZjE5LTlkYWQtYWZkOGNiZjI2NzE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEwVDEwMTgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxZGU0Nzc4ZDZmYTlhNTYwMTNhODVhMjg1Njg5ZGRkZmFmZWJjNTgxN2E1M2RkNWE0MDY0OWU2ODg0NzgzNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BeDC7VXUutG_OVkPLpkAG720Md5w4PlA60SVU9b9S-g)
> 
> Its still a work in progress but if you need help explaining any part let me know

A small change suggestion:

![image](https://github.com/wger-project/wger/assets/1658570/dadb5448-872f-4446-b87d-f0b41a9d6194)


The 'workouts' should not be coupled to a 'day of the week'.

**Comment by rolandgeider** [comment]
Hey all! I want to finally invite you to take a look at the flexible routines! 🥳 While this is not yet release ready, everything should work. The main changes are:

* (better? and) more interactive edit mode. Allow to specify ranged values, i.e. 4-6 reps
routines don't need to fit into one week. You can add as many days as needed and they will repeat in a loop, eg. A - rest - B - rest - A - rest - etc.
* progression rules: if you want, you can add progression rules so that e.g. you increase the weight by 2.5 kg every week till the fifth one, where you reset to some other value. These can get a condition so that they only increase if you actually did reach the weight in the past. You obviously will need to log the weights in order for this to work, though.
* statistics: there are a couple of new statistics pages
* lots of internal cleanups (you probably don't really care about this one)

If there is interest, I can also push a docker image.

https://wger-routines.rge.uber.space
Username: user
Password: flutteruser
(also know that this server is running on duct tape and prayers and will be somewhat slow)

![Image](https://github.com/user-attachments/assets/8606118f-1fdf-4726-abe0-42a5876fab42)

![Image](https://github.com/user-attachments/assets/f232a770-26c2-4386-af3c-a11f976692e8)

![Image](https://github.com/user-attachments/assets/a3cc6929-2497-4b22-9e3a-4376b5cd187e)

**Comment by rolandgeider** [comment]
@BeatLink here you go: just use `wger/server:routines-beta` instead of latest in the docker compose setup. If you have an existing db, it will migrate everything automatically for you, **but you can't go back!**

**Comment by BeatLink** [comment]
Thanks!

**Comment by dometto** [comment]
Amazing!

**Comment by laur89** [comment]
Good stuff.
@rolandgeider is there a rough ETA when this might get merged?

**Comment by rolandgeider** [comment]
I just fixed the last (known) blocker, so in theory we're done but I still want to check that e.g. the db is really correctly migrated as well as make sure the mobile app works well. So I'd say one or two weeks?

@dometto yes, any comments are greatly appreciated (can also be things for the future)

**Comment by rolandgeider** [comment]
This is finally live!!
